syntax = "proto3";

package pbb;

enum Status {
    UNKNOWN = 0;
    OK = 1;
    ERROR = 2;
}

message BaseResponse {
    Status status = 1;
    string message = 2;
}

// Server<>Registry
message RegisterRequest {
    string address = 2;
}

message RegisterResponse {
    Status status = 1;
    string message = 2;
    string primary_address = 3;
    int32 replica_id = 4;
}

message InformPrimaryRequest {
    string replica_address = 1;
}

// Client<>Registry
message GetReplicaListRequest {
    string name = 1;
    string address = 2;
}

message GetReplicaListResponse {
    repeated string replicas = 1;
}

// Client<>Server
message WriteRequest {
    string filename = 1;
    string content = 2;
    string uuid = 3;
    string from_address = 4;
    int64 timestamp = 5;
}

message WriteResponse {
    Status status = 1;
    string message = 2;
    string uuid = 3;
    int64 version = 4;
}

message ReadRequest {
    string uuid = 1;
}

message ReadResponse {
    Status status = 1;
    string message = 2;
    string filename = 3;
    string content = 4;
    int64 version = 5;
}

message DeleteRequest {
    string uuid = 1;
    string from_address = 2;
    int64 timestamp = 3;
}

service Registry {
    rpc Register(RegisterRequest) returns (RegisterResponse);
    rpc GetReplicaList(GetReplicaListRequest) returns (GetReplicaListResponse);
}

service Replica {
    rpc Read(ReadRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (WriteResponse);
    rpc Delete(DeleteRequest) returns (BaseResponse);

    rpc InformPrimary(InformPrimaryRequest) returns (BaseResponse);
}
